client/src/main/java/io/pravega/client/stream/impl/ReaderGroupStateManager.java

```diff
@@ -213,13 +216,19 @@ Segment findSegmentToReleaseIfRequired() {
      * the reader with the least assigned to it.
      */
     private boolean doesReaderOwnTooManySegments(ReaderGroupState state) {
-        Map<String, Double> sizesOfAssignemnts = state.getRelativeSizes();
+        // The previous calculation for this method is cached during TIME_UNIT to prevent excessive per-event computations.
```

Comment:
I am confused by this, because this method should only be invoked once per checkpoint interval. If that's set to say 20sec then this new time logic is totally redundant. Am I missing something?

Reply:
You are right, this method is being invoked much more often than it should, as I understand its goal. This method is invoked in two different places, and the one that takes a lot of CPU cycles comes from `shouldAcquireSegment()`, which as you can see, it can be potentially invoked on every read event only if `acquireTimer.hasRemaining()` is `false` most of the time, which seems to be what is going on according to the flamegraphs. The `acquireTimer` should be generally in the order of seconds, right? So maybe I can spend some more time to see if this timer is incorrectly set at some point.